CLI db migrate/rollback

vinioliveira opened the pull request:
As discussed here: #137 and #138

I haven't finished yet, still missing clean up the code as well, implement more tests. 
However, I've opened the PR to see what you guys think. 
As it's my first contribution, I tried to follow same pattern that I found, let me know if you have any guideline or style guide.


###NOTE
I did `db rollback --step=1` instead of `db rollback STEP=1`
I'd prefer Thor's options style instead of nested subcommand or environment variable. 
I think it's more explicit and it can be found on lotus CLI help.

joneslee85 commented on lib/lotus/cli.rb:
```diff
@@ -3,6 +3,38 @@
+      end
+    end
+    
+    desc 'rollback', 'roll back your migration'
```
/s/roll back/rollback

joneslee85 commented on lib/lotus/cli.rb:
```diff
@@ -3,6 +3,38 @@
+    end
+    
+    desc 'rollback', 'roll back your migration'
+    method_option :step,   desc: 'number of stpes to roll back' , type: :numeric
```
* nitpick, remove spaces in front of `desc`
* remove space after `roll back'`

joneslee85 commented on lib/lotus/cli.rb:
```diff
@@ -3,6 +3,38 @@
 
 module Lotus
+
+  class DBSubcommand < Thor 
```
What do you think about `DBCommand`?

joneslee85 commented on lib/lotus/cli.rb:
```diff
@@ -3,6 +3,38 @@
+      if options[:help] 
+        invoke :help, ['migrate']
+      else
+        Lotus::Commands::DB.new(name,environment).migrate
```
nitpick, add a space after `name,`

vinioliveira replied:
Tks, I'll fix this.

joneslee85 commented on lib/lotus/commands/db.rb:
```diff
@@ -0,0 +1,59 @@
+      def initialize(name=nil,environment)
+        @name = name
+        @environment = environment
+        @options = merge_options(@environment)
```
what do you think about instantiate a migrator object in the constructor? `@migrator = Lotus::Model::Migrator.new(adapter_config)`

joneslee85 commented on lib/lotus/commands/db.rb:
```diff
@@ -0,0 +1,59 @@
+      end
+
+      def step
+        options.fetch(:step) 
```
what do you think about setting default value for step here instead? that is `options.fetch(:step, DEFAULT_STEP)`

joneslee85 commented on lotusrb.gemspec:
```diff
@@ -22,6 +22,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'lotus-utils',      '~> 0.3', '>= 0.3.4'
   spec.add_dependency 'lotus-router',     '~> 0.2', '>= 0.2.1'
   spec.add_dependency 'lotus-controller', '~> 0.3', '>= 0.3.2'
+  spec.add_dependency 'lotus-model',      '~> 0.2.2', '>= 0.2.2'
```
Can you please add point lotus-model to my migration branch for the time being? It would help you do integration test. FYI, you have to specify it in Gemfile instead, gemspec does not allow pointing to github repo.

vinioliveira replied:
Sure, I should've not committed this file because I don't know what put on it.
I'm using local reference on my Gemfile but I didn't commit because it's referring local repo. I'll change both to point to your branch.

joneslee85 commented on lib/lotus/commands/db.rb:
```diff
@@ -0,0 +1,59 @@
+      attr_reader :environment, :options, :name
+      DEFAULTS = { step:1 }
+
+      def initialize(name=nil,environment)
```
nitpick, add space in between params

joneslee85 commented on lib/lotus/commands/db.rb:
```diff
@@ -0,0 +1,59 @@
+module Lotus
+  module Commands
+    class DB
+      attr_reader :environment, :options, :name
```
What do you think about making these readers private? I don't see any reason behind exposing it to the outside world at this stage.

vinioliveira replied:
I left all theses readers for test purpose, while I think how I'll test this.

joneslee85 replied:
I'd prefer a integration test, you could have a fixture app that has db/migrations and apps/web/db/migrations and run migration with Ruby's `system` method then check to see if migration is applied correctly, please see my integration test on lotus/model branch for ideas.

joneslee85 commented:
Looks good, now all we need is integration tests :+1:

AlfonsoUceda commented:
and document classes and methods please ;)

vinioliveira commented:
They are on their way :+1:

vinioliveira commented on lib/lotus/commands/db.rb:
```diff
@@ -0,0 +1,69 @@
+
+      DEFAULT_STEP = 1 
+      APP_MIGRATION_DIRECOTRY = "apps/web/db/migrations".freeze
+      MIGRATION_DIRECTORY = "db/migrations".freeze
```
I could call Lotus::Model::Migration:: MIGRATION_DIRECTORY in this case I think Lotus::Model::Migration should have Lotus::Model::Migration::APP_MIGRATION_DIRECTORY as well.

joneslee85 replied:
@vinioliveira I don't think Lotus::Model::Migration should be aware of the directories defined by Lotus (lotusrb). Firstly, Lotus::Model is designed to be framework agnostic. Secondly, I don't think Lotus::Model is the right place to hold this value.

vinioliveira commented:
I'll document class and method tomorrow morning.

joneslee85 commented on lib/lotus/commands/db.rb:
```diff
@@ -0,0 +1,69 @@
+
+      def migration_directory
+        if @name 
+          pwd.join(APP_MIGRATION_DIRECOTRY)
```
we can't hardcode directory. A lotus app can have many apps, the `web` is just a default app that is created, what if user want to run it against other apps? For the first phase of this feature, can you please remove the support for running migration against apps? I want to get migration done for the lib/app folder done correctly. Then we could discuss how we go about implementing the migration for apps, I am thinking of syntax `bundle exec lotus db migrate --app=web`

vinioliveira replied:
Sure, I didn't realised that. What do you mean by "remove the support" ? Should I remove all the implementation or just remove from the CLI is fine ?  
BTW the currently implementation follows the given syntax  `bundle exec lotus db migrate web`

joneslee85 commented on test/integration/db_migrations_test.rb:
```diff
@@ -0,0 +1,105 @@
+      db_command.migrate   
+    end
+
+    describe 'when not pass lotus app' do
```
What do you think about rewording this sentence to `when app param is not specified`?

joneslee85 commented on lib/lotus/commands/db.rb:
```diff
@@ -0,0 +1,69 @@
+        @environment  = environment
+        @options      = @environment.to_options
+        load_environment
+        @migrator     = Lotus::Model::Migrator.new(adapter_config)
```
Can you please set `logger: Lotus::Logger.new`? In our lotus/model constructor, I use `::Logger`, however we should take advantage of the lotus logger here instead. Please see reference (https://github.com/lotus/model/pull/144/files#diff-baefffc88152883d05db7211163864e0R126)

joneslee85 commented:
@vinioliveira it's about time to rebase and to merge! Yay!

vinioliveira commented:
@joneslee85 :+1: sorry for the delay. I'll review your branch to make any necessary modifications on my implementation.

joneslee85 commented:
@vinioliveira ping

vinioliveira commented:
@joneslee85 sorry for the delay, I've moved to other city and started a new job, so I ran out of time. 

Just to clarify things, we're going to release in this version the app options for migrations ? I implemented in the past but I've removed. So just let me know if we going to implement this, then, I can provide this feature. 

I have updated my branch with the latest version your branch, things seems to be ok.

joneslee85 commented:
I close this issue because @jodosha has implemented this feature in his recent work already. Thanks much for helping.
